File: scratchpad/foo_thread.rb
# Demo application showing how once can combine the Ruby
# threading module with GObject signals to make a simple thread
# manager class which can be used to stop horrible blocking GUIs.
#
# (c) 2009, ported to Ruby by Daniel Lucraft <dan@fluentradical.com>
# (c) 2008, John Stowers <john.stowers@gmail.com>
#
# This program serves as an example, and can be freely used, copied, derived
# and redistributed by anyone. No warranty is implied or given.

require 'gtk2'

# Cancellable thread which uses gobject signals to return information
# to the GUI.
class FooThread < GLib::Object
  type_register
  
  signal_new("completed",    # name
    GLib::Signal::RUN_LAST, # flags
    nil,                     # accumulator (XXX: not supported yet)
    nil                     # return type (void == nil) 
      # parameter types
    )
      
  signal_new("progress",    # name
    GLib::Signal::RUN_LAST, # flags
    nil,                     # accumulator (XXX: not supported yet)
    nil,                     # return type (void == nil) 
    Float # parameter types
  )
  
  def initialize(data, name)
    @data, @name = args[0], args[1]
  end
  
  def run
    print "Running #{self}"
    0.upto(data) do |i|
      sleep 0.1
      emit("progress", i.to_f/data*100)
    end
    emit("completed")
  end
  
  # Override gobject.GObject to always emit signals in the main thread
  # by emmitting on an idle handler
  def emit(*args)
    Gtk.idle_add do
      super
    end
  end
end

# Manages many FooThreads. This involves starting and stopping
# said threads, and respecting a maximum num of concurrent threads limit
class FooThreadManager
  def initialize(max_concurrent_threads)
    @max_concurrent_threads = max_concurrent_threads
    @foo_threads = {}
    @threads = []
    @pending_foo_thread_args = []
  end

  # Decrements the count of concurrent threads and starts any 
  # pending threads if there is space
  def register_thread_completed(foo_thread, *args)
    @foo_threads.delete(args)
    running = @foo_threads.length - @pending_foo_thread_args.length
    pending = @pending_foo_thread_args.length
    puts "#{foo_thread} completed. #{running} running, #{pending} pending"
    
    if running < @max_concurrent_threads
      args = @pending_foo_thread_args.shift
      puts "Starting pending #{args.inspect}"
      @threads << Thread.new { @foo_threads[args].run }
    end
  end

  # Makes a thread with args. The thread will be started when there is
  # a free slot
  def make_thread(completed_cb, progresscb, user_data, *args)
    running = @foo_threads.length - @pending_foo_thread_args.length
    
    unless @foo_threads.include? args
      foo_thread = FooThread.new(*args)
      foo_thread.signal_connect("completed", user_data, &completed_cb)
      foo_thread.signal_connect("completed") do ||
        _register_thread_completed(foo_thread, *args)
      end
      foo_thread.signal_connect("progress", user_data, &progress_cb)
      @foo_threads[args] = foo_thread
      
      if running < @max_concurrent_threads
        puts "Starting #{foo_thread}"
        @threads << Thread.new { @foo_threads[args].run }
      else
        puts "Queuing #{foo_thread}"
        @pending_foo_thread_args.append(args)
      end
    end
  end

  # Stops all threads. If block is True then actually wait for the thread
  # to finish (may block the UI) 
  def stop_all_threads(block=false)
    @threads.each do |thread|
      thread.terminate
    end
  end
end

class Demo
  def initialize
end

Gdk::Threads.init

File: scratchpad/gtk-stuff/FooThread.py
# Demo application showing how once can combine the python
# threading module with GObject signals to make a simple thread
# manager class which can be used to stop horrible blocking GUIs.
#
# (c) 2008, John Stowers <john.stowers@gmail.com>
#
# This program serves as an example, and can be freely used, copied, derived
# and redistributed by anyone. No warranty is implied or given.
import gtk
import gobject
gobject.threads_init()

import threading
import time
import random

class _IdleObject(gobject.GObject):
    """
    Override gobject.GObject to always emit signals in the main thread
    by emmitting on an idle handler
    """
    def __init__(self):
        gobject.GObject.__init__(self)

    def emit(self, *args):
        gobject.idle_add(gobject.GObject.emit,self,*args)

class _FooThread(threading.Thread, _IdleObject):
    """
    Cancellable thread which uses gobject signals to return information
    to the GUI.
    """
    __gsignals__ =  { 
            "completed": (
                gobject.SIGNAL_RUN_LAST, gobject.TYPE_NONE, []),
            "progress": (
                gobject.SIGNAL_RUN_LAST, gobject.TYPE_NONE, [
                gobject.TYPE_FLOAT])        #percent complete
            }

    def __init__(self, *args):
        threading.Thread.__init__(self)
        _IdleObject.__init__(self)
        self.cancelled = False
        self.data = args[0]
        self.name = args[1]
        self.setName("%s" % self.name)

    def cancel(self):
        """
        Threads in python are not cancellable, so we implement our own
        cancellation logic
        """
        self.cancelled = True

    def run(self):
        print "Running %s" % str(self)
        for i in range(self.data):
            if self.cancelled:
                break
            time.sleep(0.1)
            self.emit("progress", i/float(self.data)*100)            
        self.emit("completed")

class FooThreadManager:
    """
    Manages many FooThreads. This involves starting and stopping
    said threads, and respecting a maximum num of concurrent threads limit
    """
    def __init__(self, maxConcurrentThreads):
        self.maxConcurrentThreads = maxConcurrentThreads
        #stores all threads, running or stopped
        self.fooThreads = {}
        #the pending thread args are used as an index for the stopped threads
        self.pendingFooThreadArgs = []

    def _register_thread_completed(self, thread, *args):
        """
        Decrements the count of concurrent threads and starts any 
        pending threads if there is space
        """
        del(self.fooThreads[args])
        running = len(self.fooThreads) - len(self.pendingFooThreadArgs)

        print "%s completed. %s running, %s pending" % (
                            thread, running, len(self.pendingFooThreadArgs))

        if running < self.maxConcurrentThreads:
            try:
                args = self.pendingFooThreadArgs.pop()
                print "Starting pending %s" % self.fooThreads[args]
                self.fooThreads[args].start()
            except IndexError: pass

    def make_thread(self, completedCb, progressCb, userData,  *args):
        """
        Makes a thread with args. The thread will be started when there is
        a free slot
        """
        running = len(self.fooThreads) - len(self.pendingFooThreadArgs)

        if args not in self.fooThreads:
            thread = _FooThread(*args)
            #signals run in the order connected. Connect the user completed
            #callback first incase they wish to do something 
            #before we delete the thread
            thread.connect("completed", completedCb, userData)
            thread.connect("completed", self._register_thread_completed, *args)
            thread.connect("progress", progressCb, userData)
            #This is why we use args, not kwargs, because args are hashable
            self.fooThreads[args] = thread

            if running < self.maxConcurrentThreads:
                print "Starting %s" % thread
                self.fooThreads[args].start()
            else:
                print "Queing %s" % thread
                self.pendingFooThreadArgs.append(args)

    def stop_all_threads(self, block=False):
        """
        Stops all threads. If block is True then actually wait for the thread
        to finish (may block the UI) 
        """
        for thread in self.fooThreads.values():
            thread.cancel()
            if block:
                if thread.isAlive():
                    thread.join()

class Demo:
    def __init__(self):
        #build the GUI
        win = gtk.Window()
        win.connect("delete_event", self.quit)
        box = gtk.VBox(False,4)
        win.add(box)
        addButton = gtk.Button("Add Thread")
        addButton.connect("clicked",self.add_thread)
        box.pack_start(addButton,False,False)
        stopButton = gtk.Button("Stop All Threads")
        stopButton.connect("clicked",self.stop_threads)
        box.pack_start(stopButton,False,False)

        #display threads in a treeview
        self.pendingModel = gtk.ListStore(gobject.TYPE_STRING,gobject.TYPE_INT)
        self.completeModel = gtk.ListStore(gobject.TYPE_STRING)
        self._make_view(self.pendingModel, "Pending Threads", True, box)
        self._make_view(self.completeModel,"Completed Threads", False, box)

        #THE ACTUAL THREAD BIT
        self.manager = FooThreadManager(3)

        #Start the demo
        win.show_all()
        gtk.main()

    def _make_view(self, model, title, showProgress, vbox):
        view = gtk.TreeView(model)
        view.append_column(gtk.TreeViewColumn(title, gtk.CellRendererText(), text=0))
        if showProgress:
            view.append_column(gtk.TreeViewColumn("Progress", gtk.CellRendererProgress(), value=1))
        vbox.pack_start(view)

    def quit(self, sender, event):
        self.manager.stop_all_threads(block=True)
        gtk.main_quit()

    def stop_threads(self, *args):
        #THE ACTUAL THREAD BIT
        self.manager.stop_all_threads()

    def add_thread(self, sender):
        #make a thread and start it
        data = random.randint(20,60)
        name = "Thread #%s" % random.randint(0,1000)
        rowref = self.pendingModel.insert(0,(name,0)) 
        
        #THE ACTUAL THREAD BIT
        self.manager.make_thread(
                        self.thread_finished,
                        self.thread_progress,
                        rowref,data,name)

    def thread_finished(self, thread, rowref):
        self.pendingModel.remove(rowref)
        self.completeModel.insert(0,(thread.name,))

    def thread_progress(self, thread, progress, rowref):
        self.pendingModel.set_value(rowref,1,int(progress))

if __name__ == "__main__":
    demo = Demo()
